Import Libraries

In [1]:
import pandas as pd
import numpy as np
import os

Create Pathname

In [2]:
path=r'/Users/evancarr/Desktop/CF Exercises/Instacart Basket Analysis'

Import Orders.csv, Products.csv, and Departments.csv from Prepared Data

In [3]:
#Import orders.csv
df_ords=pd.read_csv(os.path.join(path, '02 Data','Prepared Data','orders_wrangled.csv'), index_col=False)

In [7]:
#Import Products.csv
df_prods = pd.read_csv(os.path.join(path,'02 Data', 'Original Data', 'products.csv'), index_col = False)

Question 01.
Performing consistency checks on products.csv and cleaning up dataframe

In [12]:
df_prods.shape

(49693, 5)

In [9]:
#checking for null values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

Creating product dataframe with no missing values

In [10]:
#creating dataframe with no missing values
df_prods_clean=df_prods[df_prods['product_name'].isnull()==False]

In [15]:
df_prods_clean.shape

(49677, 5)

Creating products dataframe with no duplicates and no missing values

In [13]:
#dropping duplicates from products.csv and renaming dataframe to df_perf_prods
df_perf_prods=df_prods_clean.drop_duplicates()

In [14]:
df_perf_prods.shape

(49672, 5)

Question 02

In [21]:
df_ords.head(33)

Unnamed: 0  order_id  customer_number  order_number  orders_day_of_week  \
0            0   2539329                1             1                   2   
1            1   2398795                1             2                   3   
2            2    473747                1             3                   3   
3            3   2254736                1             4                   4   
4            4    431534                1             5                   4   
5            5   3367565                1             6                   2   
6            6    550135                1             7                   1   
7            7   3108588                1             8                   1   
8            8   2295261                1             9                   1   
9            9   2550362                1            10                   4   
10          10   1187899                1            11                   4   
11          11   2168274                2             1                   2   
12          12   1501582                2             2                   5   
13          13   1901567                2             3                   1   
14          14    738281                2             4                   2   
15          15   1673511                2             5                   3   
16          16   1199898                2             6                   2   
17          17   3194192                2             7                   2   
18          18    788338                2             8                   1   
19          19   1718559                2             9                   2   
20          20   1447487                2            10                   1   
21          21   1402090                2            11                   1   
22          22   3186735                2            12                   1   
23          23   3268552                2            13                   4   
24          24    839880                2            14                   3   
25          25   1492625                2            15                   1   
26          26   1374495                3             1                   1   
27          27    444309                3             2                   3   
28          28   3002854                3             3                   3   
29          29   2037211                3             4                   2   
30          30   2710558                3             5                   0   
31          31   1972919                3             6                   0   
32          32   1839752                3             7                   0   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
1                   7                    15.0  
2                  12                    21.0  
3                   7                    29.0  
4                  15                    28.0  
5                   7                    19.0  
6                   9                    20.0  
7                  14                    14.0  
8                  16                     0.0  
9                   8                    30.0  
10                  8                    14.0  
11                 11                     NaN  
12                 10                    10.0  
13                 10                     3.0  
14                 10                     8.0  
15                 11                     8.0  
16                  9                    13.0  
17                 12                    14.0  
18                 15                    27.0  
19                  9                     8.0  
20                 11                     6.0  
21                 10                    30.0  
22                  9                    28.0  
23                 11                    30.0  
24                 10                    13.0  
25                 11                    30.0  
26                 14

In [16]:
df_ords.describe()

Unnamed: 0      order_id  customer_number  order_number  \
count  3.421083e+06  3.421083e+06     3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06     1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05     5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00     1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05     5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06     1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06     1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06     2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

-The 2nd column (Unnamed:0) is redundant and should be excluded from the dataframe.

-The max on 'order_number' column seems to be lower than the third quartile which should be investigated *

-The max on 'days_since_prior_order' column is also less than the third quartile *

-order_id should be turned into string data type

*after studying the table longer i noticed that the e+02 and e+01 at the end of the number represent the number of decimal spaces that need to move to the right in order to get the correct outputs ('order_number' max = 100 and 'days_since_prior_order' max = 30)

In [41]:
#dropping 'Unnamed: 0' coulumn
df_ords=df_ords.drop(columns=['Unnamed: 0'])

In [42]:
df_ords.head()

order_id  customer_number  order_number  orders_day_of_week  \
0   2539329                1             1                   2   
1   2398795                1             2                   3   
2    473747                1             3                   3   
3   2254736                1             4                   4   
4    431534                1             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [43]:
#changing 'order_id' data type to string
df_ords['order_id']=df_ords['order_id'].astype('str')

In [46]:
df_ords.dtypes

order_id                   object
customer_number             int64
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

Question 03

Testing for mixed data types in products.csv

In [29]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

/tmp/ipykernel_6058/1946670780.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
/tmp/ipykernel_6058/1946670780.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
/tmp/ipykernel_6058/1946670780.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
/tmp/ipykernel_6058/1946670780.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
/tmp/ipykernel_6058/1946670780.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].app

Question 4

No mixed type data found

Question 5

Checking for missing values

In [30]:
#checking for missing values in orders.csv
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
customer_number                0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

-days_since_prior_order has 206209 NaN values

-There are many missing vaules in this column because the customers have not placed any more orders recently. A sub dataframe could be created indicating non-recurring customers by grouping all 206209 missing values into a subset. 

Question 6

Addressing the missing values

In [33]:
#creating subset of orders data frame that includes all non-recurring orders
df_single_purchases=df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [34]:
df_single_purchases.head()

Unnamed: 0  order_id  customer_number  order_number  orders_day_of_week  \
0            0   2539329                1             1                   2   
11          11   2168274                2             1                   2   
26          26   1374495                3             1                   1   
39          39   3343014                4             1                   6   
45          45   2717275                5             1                   3   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
11                 11                     NaN  
26                 14                     NaN  
39                 11                     NaN  
45                 12                     NaN

By creating a subset of orders that include that include missing values for the 'days_since_prior_order' column we can conclude that these orders are one time purchases and therefore not recurring 

Question 7



In [37]:
#checking for duplicats in orders.csv by creating a subset
df_ord_dups=df_ords[df_ords.duplicated()]


In [38]:
df_ord_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, customer_number, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

Question 8

There were no duplicates in the orders.csv

Question 9

Export df_prods & df_ords into 'Prepared Data'


In [48]:
#exporting df_prods to Prepared Data
df_prods.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_perfected.csv'))

In [49]:
#exporting df_ords to Prepared Data
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_perfected.csv'))